nixos/modules/services/networking/suricata/default.nix

```diff
@@ -0,0 +1,218 @@
+{ config
+, pkgs
+, lib
+, ...
+}: let
+  cfg = config.services.suricata;
+  pkg = cfg.package;
+  yaml = pkgs.formats.yaml { };
+  inherit (lib) mkOption types;
+in {
+  meta.maintainers = with lib.maintainers; [ felbinger ];
+
+  options.services.suricata = {
+    enable = lib.mkEnableOption "Suricata";
+
+    package = lib.mkPackageOption pkgs "suricata" { };
+
+    configFile = mkOption {
+      type = types.path;
+      default = pkgs.writeTextFile {
+        name = "suricata.yaml";
+        text = ''
+          %YAML 1.1
+          ---
+          ${builtins.readFile (yaml.generate "suricata-settings-raw.yaml" (lib.filterAttrsRecursive (name: value: value != null) cfg.settings))}
+        '';
+      };
+      description = ''
+        The suricata configuration file, usually you don't want to override the default value
+        and just use the settings options. If you want to include extra configuration to the file
+        use the include option.
+      '';
+    };
+
+    settings = mkOption {
+      type = types.submodule (import ./settings.nix { inherit config lib yaml; });
+      example = lib.literalExpression ''
+        vars.address-groups.HOME_NET = "192.168.178.0/24";
+        outputs = [
+          {
+            fast = {
+              enabled = true;
+              filename = "fast.log";
+              append = "yes";
+            };
+          }
+          {
+            eve-log = {
+              enabled = true;
+              filetype = "regular";
+              filename = "eve.json";
+              community-id = true;
+              types = [
+                {
+                  alert.tagged-packets = "yes";
+                }
+              ];
+            };
+          }
+        ];
+        af-packet = [
+          {
+            interface = "enp0s3";
+            cluster-id = "99";
+            cluster-type = "cluster_flow";
+            defrag = "yes";
+          }
+          {
+            interface = "default";
+          }
+        ];
+        app-layer.protocols = {
+          telnet.enabled = "yes";
+          dnp3.enabled = "yes";
+          modbus.enabled = "yes";
+        };
+      '';
+      description = "Suricata settings";
+    };
+
+    enabledSources = mkOption {
+      type = types.listOf types.str;
+      default = [
+        "et/open"
+        "etnetera/aggressive"
+        "stamus/lateral"
+        "oisf/trafficid"
+        "tgreen/hunting"
+        "sslbl/ja3-fingerprints"
+        "sslbl/ssl-fp-blacklist"
+        "malsilo/win-malware"
+      ];
+      description = ''
+        List of sources that should be enabled.
+        Currently sources which require a secret-code are not supported.
+      '';
+    };
+
+    disabledRules = mkOption {
+      type = types.listOf types.str;
+      # protocol dnp3 seams to be disabled, which causes the signature evaluation to fail, so we disable the
```

Comment:
I'm not sure yet, I haven't had time to research it.

Reply:
So far I have not been able to find anything about this. Regardless of the suricata configuration when I remove these rules from the disabledRules option I get the following error:
<details>
<pre>
Jul 22 10:33:55 nixos suricata[23721]: i: suricata: This is Suricata version 7.0.5 RELEASE running in SYSTEM mode
Jul 22 10:33:55 nixos suricata[23721]: E: detect-parse: protocol "dnp3" cannot be used in a signature.  Either detection for this protocol is not yet supported OR detection has been disabled for protocol through the yaml option app-layer.protocols.dnp3.detection-enabled
Jul 22 10:33:55 nixos suricata[23721]: E: detect: error parsing signature "alert dnp3 any any -> any any (msg:"SURICATA DNP3 Request flood detected";        app-layer-event:dnp3.flooded; classtype:protocol-command-decode; sid:2270000; rev:2;)" from file /var/lib/suricata/rules/suricata.rules at line 128
Jul 22 10:33:55 nixos suricata[23721]: E: detect-parse: protocol "dnp3" cannot be used in a signature.  Either detection for this protocol is not yet supported OR detection has been disabled for protocol through the yaml option app-layer.protocols.dnp3.detection-enabled
Jul 22 10:33:55 nixos suricata[23721]: E: detect: error parsing signature "alert dnp3 any any -> any any (msg:"SURICATA DNP3 Length too small";        app-layer-event:dnp3.len_too_small; classtype:protocol-command-decode; sid:2270001; rev:3;)" from file /var/lib/suricata/rules/suricata.rules at line 129
Jul 22 10:33:55 nixos suricata[23721]: E: detect-parse: protocol "dnp3" cannot be used in a signature.  Either detection for this protocol is not yet supported OR detection has been disabled for protocol through the yaml option app-layer.protocols.dnp3.detection-enabled
Jul 22 10:33:55 nixos suricata[23721]: E: detect: error parsing signature "alert dnp3 any any -> any any (msg:"SURICATA DNP3 Bad link CRC";        app-layer-event:dnp3.bad_link_crc; classtype:protocol-command-decode; sid:2270002; rev:2;)" from file /var/lib/suricata/rules/suricata.rules at line 130
Jul 22 10:33:55 nixos suricata[23721]: E: detect-parse: protocol "dnp3" cannot be used in a signature.  Either detection for this protocol is not yet supported OR detection has been disabled for protocol through the yaml option app-layer.protocols.dnp3.detection-enabled
Jul 22 10:33:55 nixos suricata[23721]: E: detect: error parsing signature "alert dnp3 any any -> any any (msg:"SURICATA DNP3 Bad transport CRC";        app-layer-event:dnp3.bad_transport_crc; classtype:protocol-command-decode; sid:2270003; rev:2;)" from file /var/lib/suricata/rules/suricata.rules at line 131
Jul 22 10:33:55 nixos suricata[23721]: E: detect-parse: protocol "dnp3" cannot be used in a signature.  Either detection for this protocol is not yet supported OR detection has been disabled for protocol through the yaml option app-layer.protocols.dnp3.detection-enabled
Jul 22 10:33:55 nixos suricata[23721]: E: detect: error parsing signature "alert dnp3 any any -> any any (msg:"SURICATA DNP3 Unknown object";        app-layer-event:dnp3.unknown_object; classtype:protocol-command-decode; sid:2270004; rev:2;)" from file /var/lib/suricata/rules/suricata.rules at line 132
</pre>
</details>

I also couldn't find a compile option to enable/disable a protocol.

Since DNP3 is such a rare protocol that is most commonly used in industrial control systems (a place where I can't imagine using NixOS at the moment), I don't think it's necessary to implement this right now. What are your thoughts on this?